### ISSUE TITLE
Incorrect cropping in modern mode for US units

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/gmt#497]
**Description of the problem**

The following commands should generate three identical plots. However, in my computer, map1.png is not correctly cropped. map2.png and map3.png are also slightly different. See attached figures below. 

```bash
gmt begin map1 png
gmt coast -Rg -JH15c -Baf -W1p -A10000
gmt end

gmt coast -Rg -JH15c -Baf -W1p -A10000 -png map2

gmt pscoast -Rg -JH15c -Baf -W1p -A10000 > map3.ps
gmt psconvert -A -P -Tg map3.ps

```

**map1.png:**
![image](https://user-images.githubusercontent.com/3974108/54171451-b5d1aa80-4450-11e9-8dba-54cba909e00d.png)

**map2.png:**
![image](https://user-images.githubusercontent.com/3974108/54171459-bff3a900-4450-11e9-85d1-d5e8462f2340.png)

**map3.png:**
![image](https://user-images.githubusercontent.com/3974108/54171477-cbdf6b00-4450-11e9-9c43-77e5d3eeadc3.png)

**System information**

* Operating system: macOS
* Version of GMT: 6.0.0 latest
* Version of gs: 9.26, installed via homebrew

**Comment by PaulWessel** [comment]
For me, 1 and 2 are identical. 3 is different because classic and modern has different behavior as to which GSHHG data set they pick.  Classic sets **-Dl** (low) by default while modern elects the most reasonable resolution given scale.

**Comment by seisman** [comment]
psconvert debug information for 1:
```
end [DEBUG]: End Workflow.  Session Name = 50080. Directory /Users/seisman/.gmt/sessions/gmt6.50080 removed.
end [DEBUG]: No figure file /Users/seisman/.gmt/sessions/gmt6.50080/gmt.figures - nothing to do
end [DEBUG]: Got session name as map1 and default graphics formats as png
end [INFORMATION]: Process GMT figure queue: 1 figures found
end [INFORMATION]: Processing GMT figure #0 [map1 png ]
end [DEBUG]: psconvert: /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps- -Tg -Fmap1 -A
end (gmtlib_free_tmp_arrays): tried to free unallocated memory
psconvert [DEBUG]: Complete partial PS file /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-
psconvert [DEBUG]: Size of half-baked PS file = 71609.
psconvert [DEBUG]: Fattened up PS file /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-
psconvert [INFORMATION]: Processing /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-...
psconvert [INFORMATION]: Find HiResBoundingBox ...
psconvert [DEBUG]: Running: gs -q -dSAFER -dNOPAUSE -dBATCH -sDEVICE=bbox -DPSL_no_pagefill -dMaxBitmap=2147483647 -dUseFastColor=true '/Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-' 2> './psconvert_50083c.bb'
psconvert [INFORMATION]: Figure dimensions: Width: 273.006 points [3.79175 inch]  Height: 214.092 points [2.9735 inch]
psconvert [DEBUG]: Delete ./psconvert_50083c.bb
psconvert [INFORMATION]: [338.994 359.244 612 573.336]...
psconvert [INFORMATION]: Convert to PNG...
psconvert [DEBUG]: Running: gs -q -dNOPAUSE -dBATCH -dSCANCONVERTERTYPE=2 -dMaxBitmap=2147483647 -dUseFastColor=true -dGraphicsAlphaBits=2 -dTextAlphaBits=4 -sDEVICE=png16m  -g1138x893 -r300 -sOutputFile='map1.png' -f'./psconvert_50083d.eps'
psconvert [DEBUG]: Delete ./psconvert_50083d.eps
psconvert [DEBUG]: Final input buffer length was 256
psconvert (gmtlib_free_tmp_arrays): tried to free unallocated memory
```

psconvert debug information for 2:
```
end [DEBUG]: End Workflow.  Session Name = 50080. Directory /Users/seisman/.gmt/sessions/gmt6.50080 removed.
end [DEBUG]: No figure file /Users/seisman/.gmt/sessions/gmt6.50080/gmt.figures - nothing to do
end [DEBUG]: Got session name as map2 and default graphics formats as png
end [INFORMATION]: Process GMT figure queue: 1 figures found
end [INFORMATION]: Processing GMT figure #0 [map2 png ]
end [DEBUG]: psconvert: /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps- -Tg -Fmap2 -A
end (gmtlib_free_tmp_arrays): tried to free unallocated memory
psconvert [DEBUG]: Complete partial PS file /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-
psconvert [DEBUG]: Size of half-baked PS file = 71627.
psconvert [DEBUG]: Fattened up PS file /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-
psconvert [INFORMATION]: Processing /Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-...
psconvert [INFORMATION]: Find HiResBoundingBox ...
psconvert [DEBUG]: Running: gs -q -dSAFER -dNOPAUSE -dBATCH -sDEVICE=bbox -DPSL_no_pagefill -dMaxBitmap=2147483647 -dUseFastColor=true '/Users/seisman/.gmt/sessions/gmt6.50080/gmt_0.ps-' 2> './psconvert_50088c.bb'
psconvert [INFORMATION]: Figure dimensions: Width: 467.064 points [6.487 inch]  Height: 214.092 points [2.9735 inch]
psconvert [DEBUG]: Delete ./psconvert_50088c.bb
psconvert [INFORMATION]: [338.994 359.244 806.058 573.336]...
psconvert [INFORMATION]: Convert to PNG...
psconvert [DEBUG]: Running: gs -q -dNOPAUSE -dBATCH -dSCANCONVERTERTYPE=2 -dMaxBitmap=2147483647 -dUseFastColor=true -dGraphicsAlphaBits=2 -dTextAlphaBits=4 -sDEVICE=png16m  -g1947x893 -r300 -sOutputFile='map2.png' -f'./psconvert_50088d.eps'
psconvert [DEBUG]: Delete ./psconvert_50088d.eps
psconvert [DEBUG]: Final input buffer length was 256
psconvert (gmtlib_free_tmp_arrays): tried to free unallocated memory
```

**Comment by PaulWessel** [comment]
For me, debug from 1:

```
end [DEBUG]: End Workflow.  Session Name = 17660. Directory /Users/pwessel/.gmt/sessions/gmt6.17660 removed.
end [DEBUG]: No figure file /Users/pwessel/.gmt/sessions/gmt6.17660/gmt.figures - nothing to do
end [DEBUG]: Got session name as map1 and default graphics formats as png
end [INFORMATION]: Process GMT figure queue: 1 figures found
end [INFORMATION]: Processing GMT figure #0 [map1 png ]
end [DEBUG]: psconvert: /Users/pwessel/.gmt/sessions/gmt6.17660/gmt_0.ps- -Tg -Fmap1 -A
psconvert [DEBUG]: Complete partial PS file /Users/pwessel/.gmt/sessions/gmt6.17660/gmt_0.ps-
psconvert [DEBUG]: Size of half-baked PS file = 71458.
psconvert [DEBUG]: Fattened up PS file /Users/pwessel/.gmt/sessions/gmt6.17660/gmt_0.ps-
psconvert [INFORMATION]: Processing /Users/pwessel/.gmt/sessions/gmt6.17660/gmt_0.ps-...
psconvert [INFORMATION]: Find HiResBoundingBox ...
psconvert [DEBUG]: Running: gs -q -dSAFER -dNOPAUSE -dBATCH -sDEVICE=bbox -DPSL_no_pagefill -dMaxBitmap=2147483647 -dUseFastColor=true '/Users/pwessel/.gmt/sessions/gmt6.17660/gmt_0.ps-' 2> './psconvert_71955c.bb'
psconvert [INFORMATION]: Figure dimensions: Width: 467.064 points [16.477 cm]  Height: 214.092 points [7.55269 cm]
psconvert [DEBUG]: Delete ./psconvert_71955c.bb
psconvert [INFORMATION]: [338.994 359.244 806.058 573.336]...
psconvert [INFORMATION]: Convert to PNG...
psconvert [DEBUG]: Running: gs -q -dNOPAUSE -dBATCH -dSCANCONVERTERTYPE=2 -dMaxBitmap=2147483647 -dUseFastColor=true -dGraphicsAlphaBits=2 -dTextAlphaBits=4 -sDEVICE=png16m  -g1947x893 -r300 -sOutputFile='map1.png' -f'./psconvert_71955d.eps'
psconvert [DEBUG]: Delete ./psconvert_71955d.eps
psconvert [DEBUG]: Final input buffer length was 256

```

So you have a 612 in there (which is paper width of US letter 8.5 x 72 which would seem to indicate it is setting a US letter paper and then the plot goes off...

**Comment by seisman** [comment]
You're right. I set the default unit to US in my `ConfigUser.cmake`. 
I comment out `set (UNITS "US")` in my `ConfigUser.cmake` and it works now.

**Comment by PaulWessel** [comment]
But not sure why that happens since we are supposed to have huge pauperize in modern mode.

**Comment by joa-quim** [comment]
> Classic sets -Dl (low) by default while modern elects the most reasonable resolution given scale.

Shouldn't that behavior be based on GMT5 vs GMT6, instead of classic vs modern? I think both classic and modern should do the same thing here.

**Comment by PaulWessel** [comment]
Well, that would break backwards compatibility so I think not.

**Comment by joa-quim** [comment]
When doing major versions updates we are allowed to do breakages and this one would be for consistency and for a good reason (though at the end all compat breakages are for good reasons).

**Comment by PaulWessel** [comment]
Running

```
gmt set PS_MEDIA letter
gmt begin map1 png
gmt coast -Rg -JH15c -Baf -W1p -A10000
gmt end
```

illustrates the problem encountered by @seisman.  We need to protect against this somehow.  Perhaps only gmt set calls inside the session should count?  This goes against our initial design in which any existing gmt.conf settings would be honored and placed in the temporary work area.  But clearly this can have unintended side-effects.  Maybe the user will have to take action to enforce starting with whatever gmt.conf they wish to use.

**Comment by leouieda** [comment]
Yep. I think having this problem documented and how to solve it is better than trying to fix something that is not our fault.

**Comment by PaulWessel** [comment]
OK, then the solution to this would be

- Start each gmt session with the default conf settings and not inherit anything from the environment.
- Let users make changes via gmt set calls
- Let **gmt set** get an enhancement to accept another gmt.conf file to ingest? E.g., I may want to run _gmt set /home/my/favorite/gmt.conf_ and it will do its thing.
- Documentation of **gmt begin** will be updated to make this clear.

However, there is another issue: In our test we specify PostScript output which requires a paper size to be set.  We do this now via the gmtest.in script which sets PS_MEDIA letter. With the above changes we will get complaints that PS_MEDIA is not initialized I think.

**Comment by joa-quim** [comment]
Curiously it doesn't error with a one-liner

```
gmt set PS_MEDIA letter
gmt coast -Rg -JH15c -Baf -W1p -A10000 -png map2
```
Doesn't this suggest the problem can be avoidable?

**Comment by PaulWessel** [comment]
It certainly suggest the two paths are different.  The one-liner has some glue and scaffolding holding up a fake session so I will have to see why they are different.

**Comment by joa-quim** [comment]
Alternatively, we can ignore the PS_MEDIA setting when the format is other than PS.

**Comment by joa-quim** [comment]
No problem like this either
```
gmt begin
gmt figure map2 png
gmt coast -Rg -JH15c -Baf -W1p -A10000
gmt end
```

**Comment by seisman** [comment]
I tried to run following commands using two compiled GMT, one with SI units, another one with US units.

```
gmt get PS_MEDIA
gmt begin map1 png
gmt get PS_MEDIA
gmt coast -Rg -JH15c -Baf -W1p -A10000
gmt get PS_MEDIA
gmt end
```

For US units, the output is:
```
letter
letter
letter
```

For SI units, the output is 
```
a4
32767x32767
32767x32767
```

It seems GMT fails to change PS_MEDIA if GMT is compiled with US units.

**Comment by seisman** [comment]
> No problem like this either
> 
> ```
> gmt begin
> gmt figure map2 png
> gmt coast -Rg -JH15c -Baf -W1p -A10000
> gmt end
> ```

@joa-quim I tried your commands with GMT/US, it doesn't work for me.

```
gmt get PS_MEDIA
gmt begin
gmt get PS_MEDIA
gmt figure map2 png
gmt get PS_MEDIA
gmt coast -Rg -JH15c -Baf -W1p -A10000
gmt get PS_MEDIA
gmt end
gmt get PS_MEDIA
```

The output is:
```
letter
letter
letter
letter
letter
```

**Comment by joa-quim** [comment]
Sorry, it doesn't work for me either with SI. I must have mixed up the fig versions created with the different commands.

**Comment by leouieda** [comment]
So, should GMT ignore PS_MEDIA in modern mode? It's the basis for the system of shifting the figure and cropping in the end. But this only works in modern mode where there is no `.ps` to be seen.

**Comment by PaulWessel** [comment]
Reviving this thread.  If there is no gmt.conf file then it works.  But if I run
`gmt set PS_MEDIA a4 or letter
`
then it fails.  So this cannot just be a US/SI issue since for me it fails if PS_MEDIA is set to either a4 or letter in a gmt.conf before the session starts.  Of course, if I make PS_MEDIA big enough (20ix20i) then it also works.  I will debug and learn what is happening.

**Comment by PaulWessel** [comment]
The fix in PR #540 is to only honor PS_MEDIA when a PS file is requested.  For all other formats we don't want to use this dimensions but start at 32767 x 32767 points and crop.

**Comment by seisman** [comment]
It works now. Closed.
